### DISCUSSION TITLE
interface for suggesting new evaluation point(s)?

### DISCUSSION THREAD

**Discussion by mkaut** [coin-or/rbfopt#36] [Q&A]
I have an optimization problem with some dependencies between variables, such as "x[3]==0 implies x[4]..x[8]==0". Actually, the problem is such that i allows the higher x's to be positive, but then they only increase the 'cost', so the solver should be able to press them to zero - yet rbfopt needs many iterations to do so, possibly because the higher x's can be "useful" if x[3] is positive.

What I would like is to have a possibility to suggest a new evaluation points during the runtime of the solver. For example, in the objective function, if I see that x[3]==0 and the higher x's are positive, I would like to suggest the solver to check a solution with the same x[0]..x[3] and zero higher x's.

Is there some way of doing this? (I checked the documentation, but did not find any.)
And if not, is there some other way to achieve what I want?

**Comment by gnannicini** [comment]
There is no straightforward way to do so, as the optimization interface was not designed for the possibility of the user suggesting points. (Although this has been asked other times, so it would be a great feature to introduce.)

I think that the easiest way to do something along those lines would be the following:

Run the algorithm for some number n of iterations (for example, for 2-3 optimization cycles, see the cycle length).
Get back the list of all points evaluated by the algorithm, and their values.
Check the list, and evaluate all the "improved" points that you can derive from that list.
Restart the algorithm, giving the full list of evaluated points (including the new ones) as the initialization points, telling RBFOpt to skip its usual initialization strategy (there is an option for that).

Edit:
Actually, I think it may be possible to also just 'pause' the algorithm using the 'pause' argument, and then add new points with the add_node function, then run optimize() again -- without creating a new RbfoptAlgorithm object. That would allow the algorithm to keep some data learned during the search, so if it works, it is probably a better strategy.

**Reply by mkaut** [reply]
That sounds doable, thanks :-)

**Reply by mkaut** [reply]
RE Edit:
I see that the `add_node` method needs two version of the point values: `point` and `orig_point`. However, I have not found a description of the relation between the two.
In addition, I do not see a method to get the list of evaluated points from the `RbfoptAlgorithm` object (after it has been paused). Did I missed something (again)?

**Reply by gnannicini** [reply]
The points live in an original space and in a transformed space (used in the optimization).
You can transform between the two using rbfopt_utils.transform_domain(), passing the parameters stored in the RbfoptAlgorithm object -- check the code for some examples.

At the end of the optimization, or after you pause, you can get all points and values by looking at all_node_pos and all_node_val in your RbfoptAlgorithm object.

But, I am now realizing that there is an additional complication: if your problem has categorical variables, there is a nontrivial domain transformation performed at the beginning, to create an *extended* space. All points in RbfoptAlgorithm, whether they are original or not, live in this extended space. There are functions in RbfoptUtils to map things, but it adds a layer of complication. You can see examples in RbfoptAlgorithm, once again.

Apologies for the complexity of performing this task: as I said, the code wasn't designed with this possibility in mind.
